src/mono/mono/metadata/gc.c

```diff
@@ -249,6 +249,15 @@ mono_gc_run_finalize (void *obj, void *data)
 			return;
 	}
 
+	if (o->vtable->klass == mono_defaults.weakreference_class ||
+	    o->vtable->klass == mono_defaults.generic_weakreference_class) {
+		MonoWeakReference *wr = (MonoWeakReference*)o;
+		MonoGCHandle gc_handle = (MonoGCHandle)(wr->handleAndKind & ~(gsize)1);
+		mono_gchandle_free_internal (gc_handle);
+		wr->handleAndKind &= (gsize)1;
+		return;
+	}
```

Comment:
Hmm... maybe I misunderstood when the `mono_gc_run_finalize` method is run and on what thread.